### PR TITLE
[Relay] Added Merge Composite pass

### PR DESCRIPTION
**Pull request by mbarrett97**
This adds a pass, MergeComposite, which merges patterns (expressed as Relay expressions). The detected patterns are wrapped in a function call which is marked with a 'Composite' attribute that names the pattern.

This is primarily for use with the external codegen infrastructure in the case where a combination of Relay ops map to a single external codegen op. It is not the same as fusion.

Further discussion on this PR can be found at https://discuss.tvm.ai/t/rfc-external-codegen-defining-composite-relay-operators/5470.

### CODE REVIEW

File: python/tvm/relay/transform.py [hunk 1]
```diff
@@ -508,6 +508,41 @@ def Legalize(legalize_map_attr_name="FTVMLegalize"):
     return _transform.Legalize(legalize_map_attr_name)
 
 
+def AnnotateCompiler(compiler):
```
Comment (masahi) [inline python/tvm/relay/transform.py]: remove this

**Comment by masahi** [comment]
From your test cases, I don't see why this pass would help remove pattern matching on codegen side. Since Relay doesn't have `add_sub_mul` op, I still need to first find multiply, then traverse further to find sub etc, no? From the codegen side I don't see the difference compared to manual partitioning. 

It definitely removes the need for writing custom annotator for each pattern, though.

File: tests/python/relay/test_pass_merge_composite.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+        a  b
+        \ /               a  b
+        add    ====>      \ /
+         |             add_relu
+       ReLu
+
+    """
+    pattern_table = {
+        "add_sub_mul": make_add_relu_pattern()
```
Comment (masahi) [inline tests/python/relay/test_pass_merge_composite.py]: add_relu

**Comment by masahi** [comment]
ok reading your original proposal carefully I understand that I can look up 'Composite' attribute to know if a particular function has a pattern I'm looking for. Of course I still need to traverse the arguments, but I can remove the "detection" part from DetectFusedConv2DBiasReLU. 

Sounds great!

File: tests/python/relay/test_pass_merge_composite.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+        add_relu = relay.Function([in_1, in_2], relu_node)
+
+        # merged function
+        r = relay.Call(add_relu, [a, b])
+        return relay.Function([a, b], r)
+
+    result = run_opt_pass(before(), relay.transform.MergeComposite(pattern_table))
+    expected = run_opt_pass(expected(), relay.transform.InferType())
+    assert relay.analysis.alpha_equal(result, expected)
+
```
Comment (masahi) [inline tests/python/relay/test_pass_merge_composite.py]: can you add an example of retrieving the Composite attribute and verify it?

File: tests/python/relay/test_pass_merge_composite.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+        # merged function
+        add_sub_mul_1 = relay.Call(add_sub_mul, [a, b])
+        add_sub_mul_2 = relay.Call(add_sub_mul, [c, add_sub_mul_1])
+        r = relay.nn.relu(add_sub_mul_2)
+        return relay.Function([a, b, c], r)
+
+    result = run_opt_pass(before(), relay.transform.MergeComposite(pattern_table))
+    expected = run_opt_pass(expected(), relay.transform.InferType())
+    assert relay.analysis.alpha_equal(result, expected)
```
Comment (masahi) [inline tests/python/relay/test_pass_merge_composite.py]: you are missing ```if __name__ == "__main__":``` here

File: src/relay/pass/merge_composite.cc [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+ * under the License.
+ */
+
+/*!
+ * \file src/relay/pass/merge_composite.cc
+ * \brief Merges expressions matching patterns into functions marked
+ * as 'composite'.
+ */
+
+#include <tvm/top/operation.h>
```
Comment (masahi) [inline src/relay/pass/merge_composite.cc]: top -> te after rebase

**Comment by masahi** [comment]
ok I already have a working test case for detecting conv + bias + relu (Note the "Composite" attribute). Now I really want to hook this up with dnnl "compiler" attribute to enable codegen.

```
def @main(%data: Tensor[(1, 3, 224, 224), float32]) -> Tensor[(1, 16, 224, 224), float32] {
  %2 = fn (%data1: Tensor[(1, 3, 224, 224), float32], %weight: Tensor[(16, 3, 3, 3), float32], %bias: Tensor[(16, 1, 1), float32], Primitive=1, Composite="dnnl.conv_bias_relu") -> Tensor[(1, 16, 224, 224), float32] {
    %0 = nn.conv2d(%data1, %weight, padding=[1, 1], channels=16, kernel_size=[3, 3]) /* ty=Tensor[(1, 16, 224, 224), float32] */;
    %1 = add(%0, %bias) /* ty=Tensor[(1, 16, 224, 224), float32] */;
    nn.relu(%1) /* ty=Tensor[(1, 16, 224, 224), float32] */
  };
  %3 = %2(%data, meta[relay.Constant][0] /* ty=Tensor[(16, 3, 3, 3), float32] */ /* ty=Tensor[(16, 3, 3, 3), float32] */, meta[relay.Constant][1] /* ty=Tensor[(16, 1, 1), float32] */ /* ty=Tensor[(16, 1, 1), float32] */) /* ty=Tensor[(1, 16, 224, 224), float32] */;
  %6 = fn (%data2: Tensor[(1, 16, 224, 224), float32], %weight1: Tensor[(16, 16, 3, 3), float32], %bias1: Tensor[(16, 1, 1), float32], Primitive=1, Composite="dnnl.conv_bias_relu") -> Tensor[(1, 16, 224, 224), float32] {
    %4 = nn.conv2d(%data2, %weight1, padding=[1, 1], channels=16, kernel_size=[3, 3]) /* ty=Tensor[(1, 16, 224, 224), float32] */;
    %5 = add(%4, %bias1) /* ty=Tensor[(1, 16, 224, 224), float32] */;
    nn.relu(%5) /* ty=Tensor[(1, 16, 224, 224), float32] */
  };
  %6(%3, meta[relay.Constant][2] /* ty=Tensor[(16, 16, 3, 3), float32] */ /* ty=Tensor[(16, 16, 3, 3), float32] */, meta[relay.Constant][3] /* ty=Tensor[(16, 1, 1), float32] */ /* ty=Tensor[(16, 1, 1), float32] */) /* ty=Tensor[(1, 16, 224, 224), float32] */
}
```

**Comment by mbarrett97** [comment]
To hook it up end-to-end will, for now, require writing a custom annotation pass that can recognise composite functions and understand that they need to be treated in the same way as operators. I've got a generic pass that does this, but it builds on top of @zhiics annotation PR which has since been withdrawn. If we agree that this merge_composite pass is useful then I can discuss with zhiics how we want to move forward with the annotator.

File: src/relay/pass/merge_composite.cc [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+
+    Expr expr = ExprMutator::VisitExpr_(cn);
+    call = Downcast<Call>(expr);
+    if (!call->op->IsInstance<OpNode>())
+      return call;
+
+    Op op = Downcast<Op>(call->op);
+    CHECK(op.defined());
+    for (const auto& x : pattern_map_) {
+      Call pattern = Downcast<Call>(x.second);
```
Comment (masahi) [inline src/relay/pass/merge_composite.cc]: better to add
CHECK(pattern)

**Comment by masahi** [comment]
If there are multiple patterns to detetct, can a composite function detected by one pattern be used as a part of a match for another patterns?

Not sure if this is useful though. Maybe it enables breaking up a big pattern into chunks or pattern reuse.

File: src/relay/pass/merge_composite.cc [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+        new_func = FunctionSetAttr(new_func, attr::kComposite,
+                                   tir::StringImmNode::make(x.first));
+        new_func = FunctionSetAttr(new_func, attr::kPrimitive,
+            tvm::Integer(1));
+        Array<Expr> args;
+        for (const auto& free_var : free_vars) {
+          args.push_back(args_map[free_var->name_hint()][1]);
+        }
+        auto new_call = CallNode::make(new_func, args);
+        return new_call;
```
Comment (masahi) [inline src/relay/pass/merge_composite.cc]: Does this mean if there multiple patterns which can be matched to this calll, it will return the first match as determined by tvm::Map key order?

Comment (comaniac) [reply src/relay/pass/merge_composite.cc]: Same question here. If this PR does not attempt to deal with multiple pattern matching to one call, it would be better to provide a priority number for users to specify which pattern should be matched prior to others.

Comment (mbarrett97) [reply src/relay/pass/merge_composite.cc]: Yep, this is an oversight. I can fix it pretty easily to work in priority order though and shall update the PR accordingly. I'll add tests for this case.

**Comment by mbarrett97** [comment]
It's the intention that this can be called on the entire Relay graph so that it can be used to help implement a generic annotation pass (one that is aware of composite functions). That way we can define functions similar to the 'Is Supported?' mechanism in the original annotation PR (since taken down) where you could declare in Python whether an operator was supported. That could be extended to say whether a composite function is supported without having to add pattern matching code to the annotator.

The problem there is the case where composite functions do not end up in an external function after partitioning. My thinking is to have some legalize pass after the partitioning that removes the composite functions from sections of the graph not marked as external.

**Comment by comaniac** [comment]
So you want the flow to be:
```
CompositeMerge -> Annotation -> Partitioning. 
```
I agree that this would make annotation generic and straightforward, although it seems like we don't need annotation anymore if we specify all patterns including single ops. While there are lots of approaches to do so, maybe we could accept this solution first and consider the further steps. What you do think? @zhiics 

Also, please help clarify the question asked by @masahi and me about multiple matching. Thanks.

**Comment by mbarrett97** [comment]
I think we probably can get rid of the annotation pass. If we expose the logic that determines whether an operator is supported by a given external compiler to the partitioning pass, then this will also mean we don't need to define a pattern for every single operator. Just define patterns for composite operators and have a common way of expressing whether or not a given operator (composite or otherwise) is supported by an external compiler.

That would lose some of the control that could have been gained by writing a custom annotation pass, but I'm not sure what use cases would leverage that.

**Comment by comaniac** [comment]
That's exactly what I was thinking -- we can debate if we should merge annotation to this pass, or just get rid of the annotation and make this pass general enough. Anyway can be discussed in the follow-up steps.

**Comment by masahi** [comment]
do we still need partitioning if we get rid of the annotation pass? At least for simple patterns like conv + bias + relu, I can get the same composite (or partitioning, whatever) without writing a clunky custom annotator.

**Comment by mbarrett97** [comment]
It's important to emphasise the difference between partitioned subgraphs (which correspond to 'External' functions), and the merged patterns here (which correspond to 'Composite' functions).

Composite functions are a way to group together relay patterns which correspond to a particular external codegen function (eg. qnn_conv2d + bias_add + requantize maps to a single convolution function in Arm Compute Library). External functions are a way to represent a subgraph that should not go via TVM codegen but should instead go through an external compiler (eg. DNNL).

For libraries like ACL or DNNL, it happens to be the case that you can just put every supported operator/composite function in it's own external function (partitioned subgraph) and because the library calls are standalone that works well enough.

A more complex external compiler though may want to do some advanced fusion or manipulation which means it needs to see the full supported subgraph, not just a single operator/composite function at a time. It will produce some binary artifact that computes the entire subgraph rather than a particular operator within it. All that complexity should be handled by the external compiler's codegen pass, the responsibility of the partitioner is to ensure that pass receives a valid subgraph that is as large as possible while still being fully supported and introducing no dataflow problems.

In summary, I don't think this pass replaces partitioning which is essential for the 'complex compiler' case, but I think it may replace the annotation pass.

**Comment by masahi** [comment]
thanks, such "complex compiler" use cases sounds quite interesting and the difference in Composite/External there makes a lot of sense. I also understand your earlier comments better now.

**Comment by comaniac** [comment]
While @mbarrett97 has pointed out the difference between composite/annotation and partition, I just provide one more clarification: The main task for the partition pass is transforming compiler annotations in a Relay graph to external functions and it is compiler agnostic. It means we may even consider the possibility of partitioning a Relay graph to more than one external compilers.

**Comment by mbarrett97** [comment]
I've updated the PR to account for the priority ordering of patterns. You no longer specify a pattern map, but a list of tuples (pattern_name, pattern) where the order or the tuples corresponds to the order of the matching. To get this to work, I had to pass strings between Python/C++ requiring this change: https://github.com/apache/incubator-tvm/pull/4806.

File: python/tvm/relay/transform.py [hunk 8]
```diff
@@ -513,6 +513,49 @@ def Legalize(legalize_map_attr_name="FTVMLegalize"):
...
```
Comment (comaniac) [inline python/tvm/relay/transform.py]: Is this necessary for this PR?

**Comment by mbarrett97** [comment]
I've done a big clean up + simplified the code a bit and added some comments.

File: tests/python/relay/test_pass_merge_composite.py [hunk 9]
```diff
@@ -0,0 +1,439 @@
...
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""Unit tests for merge composite."""
+from tvm import expr
+from tvm import relay
+from tvm.relay.testing import run_opt_pass
+
+"""
```
Comment (anijain2305) [inline tests/python/relay/test_pass_merge_composite.py]: Do you think we should also think about ops like concat, split etc? Basically, the ops that involve TupleNode, and TupleGetItemNode.

File: tests/python/relay/test_pass_merge_composite.py [hunk 10]
```diff
@@ -0,0 +1,439 @@
...
+    expected = run_opt_pass(expected(), relay.transform.InferType())
+    assert relay.analysis.alpha_equal(result, expected)
+
+
+def test_branch_merge():
+    """Test composite function is correctly produced from branching graph.
+
+    We would expect the pattern `make_add_sub_mul_pattern` to be merged
+    into a single op `add_sub_mul`.
+
```
Comment (anijain2305) [inline tests/python/relay/test_pass_merge_composite.py]: Can we also try two parallel add_sub_mul but with same inputs? This will ensure that Relay var variables are generated correctly when used across two subgraphs.

File: tests/python/relay/test_pass_merge_composite.py [hunk 11]
```diff
@@ -0,0 +1,439 @@
...
+       a  b  a  b
+        \/    \/
+        add  sub                       a  b
+         \   /                          \/
```
Comment (anijain2305) [inline tests/python/relay/test_pass_merge_composite.py]: Can we add 1 more testcase where Var('a') = Var('c') in the primary graph? Maybe, my previous comment already covers that.

File: tests/python/relay/test_pass_merge_composite.py [hunk 12]
```diff
@@ -0,0 +1,439 @@
...
+        r = relay.multiply(add_relu_1, b)
+        return relay.Function([data, kernel, bias, a, b], r)
+
+    result = run_opt_pass(before(), relay.transform.MergeComposite(pattern_table))
+    assert not relay.analysis.free_vars(result)
+    expected = run_opt_pass(expected(), relay.transform.InferType())
+    assert relay.analysis.alpha_equal(result, expected)
+
+
+def test_merge_order():
```
Comment (anijain2305) [inline tests/python/relay/test_pass_merge_composite.py]: What happens when pattern1 is a subset of pattern2? Basically you want to fuse as much as possible. So, in ideal cases, pattern 2. But, if not pattern 2 isn't applicable, we want to still fuse pattern 1.

File: tests/python/relay/test_pass_merge_composite.py (hunk 11 above)
Comment (anijain2305) [reply tests/python/relay/test_pass_merge_composite.py]: Sorry, 1 more testcase request - Have multiple patterns in the graph, such that, the last subgraph input variables come from multiple prior subgraphs (instead of just 1).

File: python/tvm/relay/transform.py [hunk 13]
```diff
@@ -513,6 +513,31 @@ def Legalize(legalize_map_attr_name="FTVMLegalize"):
     return _transform.Legalize(legalize_map_attr_name)
 
 
+def MergeComposite(pattern_table):
+    """Merge multiple operators into a single composite relay function.
+
+    Parameters
+    ----------
+    pattern_table : list(tuple)
+        A list of (pattern_name, pattern) tuples.
```
[the remark this answers is not in the document]
Comment (comaniac) [reply python/tvm/relay/transform.py]: Because the patterns have priorities and the index of this list implies the order. There was a discussion about how to deal with multiple patterns match to the same sub graph, and this is the workaround for this PR.

**Comment by mbarrett97** [comment]
@zhiics Regarding the add-sub/sub-add case, yes this would require two patterns with the order or merging controlled by their priority. I can't think of any general way to express both these cases as a single pattern, but if you have any thoughts I'd be glad to hear them. There is potentially an issue with requiring lots of patterns and if we can come up with some concrete examples where that may be the case then I can try and reason about how to improve the pattern matching.

File: tests/python/relay/test_pass_merge_composite.py (hunk 9 above)
Comment (mbarrett97) [reply tests/python/relay/test_pass_merge_composite.py]: To be more general, yes we will need to think about this case. For this 1st iteration I've just considered patterns which are composed of Calls. I'd prefer to start with this and add the additional functionality as and when it's required.

File: tests/python/relay/test_pass_merge_composite.py (hunk 12 above)
Comment (mbarrett97) [reply tests/python/relay/test_pass_merge_composite.py]: Have pattern 2 be above pattern 1 in the priority order. That way it will first try to fuse pattern 2 in the graph and fail, then try and fuse pattern 1 and succeed.

File: src/relay/pass/merge_composite.cc [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+};
+
+Expr MergeComposite(const Expr& expr,
+    const Array<tir::StringImm>& pattern_names, const Array<Expr>& patterns) {
+  CHECK(pattern_names.size() == patterns.size());
+  Expr merged_expr = expr;
+  // merge the patterns one-by-one in order
+  for (size_t i = 0; i < patterns.size(); i++) {
+    std::string pattern_name = pattern_names[i]->value;
+    Expr pattern = patterns[i];
```
[the remark this answers is not in the document]
Comment (mbarrett97) [reply src/relay/pass/merge_composite.cc]: I don't think it's necessary here. I'm iterating over 'i' rather than the patterns directly as I'm traversing two arrays, patterns and pattern_names.

File: tests/python/relay/test_pass_merge_composite.py [hunk 15]
```diff
@@ -0,0 +1,439 @@
...
+    # check C highest priority
+    pattern_table = [
+        ("C", pattern_A()),
+        ("A", pattern_B()),
+        ("B", pattern_C()),
+    ]
+    result = run_opt_pass(before(), relay.transform.MergeComposite(pattern_table))
+    assert not relay.analysis.free_vars(result)
+    expected = run_opt_pass(after_A_priority(), relay.transform.InferType())
+    assert relay.analysis.alpha_equal(result, expected)
```
[the remark this answers is not in the document]
Comment (mbarrett97) [reply tests/python/relay/test_pass_merge_composite.py]: It's not particularly meaningful to compile beyond this point as it requires a specific codegen pass to be able to interpret the composite function. I think that coverage will come from the tests for external codegens that make use of this infrastructure.

File: tests/python/relay/test_pass_merge_composite.py (hunk 3 above)
Comment (mbarrett97) [reply tests/python/relay/test_pass_merge_composite.py]: Unfortunately this doesn't seem to be exposed by the Python API (I can set attributes but not retrieve them).

**Comment by zhiics** [comment]
@mbarrett97 Considering that this is a relatively standalone pass and it helps fusion with external functions, I think it is okay to take it in.

For the pattern matching cases, one possible way I am thinking of is automatically generating some possible pattern given some metadata.

@comaniac and I will start an RFC to discuss how we can add the whitelist based annotation back. It probably will leverage the pattern matching here.

@masahi @anijain2305 PTAL and share your comments if there is any. Thanks.

**Comment by anijain2305** [comment]
My only request is to add more test cases. In my experience, things get very ugly as the networks get bigger and the things that seems corner cases become very common. But, I am ok with the scope of this PR and delaying aggressive testing to later PR. I think it helps in quick flush of e2e pileline.

**Comment by anijain2305** [comment]
Oh just saw. The tests are already added. Thanks, its good from my side :)

**Comment by masahi** [comment]
@zhiics @comaniac It is worth discussing if we can use composite and partitioning passes to remove the annotation pass, as mentioned by @mbarrett97

**Comment by masahi** [comment]
Thanks @mbarrett97 @comaniac @zhiics @anijain2305 this is merged.

**Comment by mbarrett97** [comment]
@masahi @comaniac @zhiics Thanks for the reviews. An RFC on alternative annotation mechanisms would be great.

File: tests/python/relay/test_pass_merge_composite.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (masahi) [reply tests/python/relay/test_pass_merge_composite.py]: my DNNL PR should do this job.

File: tests/python/relay/test_pass_merge_composite.py (hunk 3 above)
Comment (jwfromm) [reply tests/python/relay/test_pass_merge_composite.py]: Whats stopping Composite from being exposed? It seems pretty important to be able to see the composite name in python for a lot of optimizations.

Comment (mbarrett97) [reply tests/python/relay/test_pass_merge_composite.py]: It should be as simple as implementing get_attribute in Python for relay.Functions. It's not been an essential part of this PR because the majority of passes (I think all non-prototype passes?) that use this are in C++ where you can get attributes.

Comment (jwfromm) [reply tests/python/relay/test_pass_merge_composite.py]: Ok thanks, maybe I'll take a look at a separate PR that adds that functionality.
